macaroonbakery/discharge.py

```diff
@@ -0,0 +1,193 @@
+    acquire each discharge macaroon.
+    The local_key parameter may optionally hold the key of the client, in
+    which case it will be used to discharge any third party caveats with the
+    special location "local". In this case, the caveat itself must be "true".
+    This can be used be a server to ask a client to prove ownership of the
+    private key.
+    It returns a list of macaroon with m as the first element, followed by all
+    the discharge macaroons.
+    All the discharge macaroons will be bound to the primary macaroon.
+    The get_discharge function is passed a context (AuthContext),
+    the caveat(Caveat) to be discharged and encrypted_caveat (bytes)will be
+    passed the external caveat payload found in m, if any.
+    '''
+    primary = m.macaroon()
+    discharges = [primary]
+
+    # cav holds the macaroon caveat that needs discharge.
+    # encrypted_caveat (bytes) holds encrypted caveat if it was held
+    # externally.
+    _NeedCaveat = namedtuple('_NeedCaveat', 'cav encrypted_caveat')
+    need = []
+
+    def add_caveats(m):
+        for cav in m.macaroon().caveats:
+            if cav.location is None or cav.location == '':
+                continue
+            key = None
+            if cav.caveat_id is not None:
+                key = cav.caveat_id.decode('utf-8')
+            encrypted_caveat = m.caveat_data()[key]
+            need.append(
+                _NeedCaveat(cav=cav,
+                            encrypted_caveat=encrypted_caveat))
+    add_caveats(m)
+    while len(need) > 0:
+        cav = need[0]
+        need = need[1:]
+        if local_key is not None and cav.cav.location == 'local':
+            # TODO use a small caveat id.
+            dm, err = discharge(ctx=ctx, key=local_key,
+                                checker=_LocalDischargeChecker(),
+                                caveat=cav.encrypted_caveat,
+                                id=cav.cav.Id,
+                                locator=_EmptyLocator())
+        else:
+            dm = get_discharge(ctx, cav.cav, cav.encrypted_caveat)
+        # It doesn't matter that we're invalidating dm here because we're
+        # about to throw it away.
+        discharge_m = dm.macaroon()
+        m = primary.prepare_for_request(discharge_m)
+        discharges.append(m)
+        add_caveats(dm)
+    return discharges
+
+
+class ThirdPartyCaveatChecker(object):
+    ''' Holds a function that checks third party caveats for validity.
+
+    If the caveat is valid, it returns a nil error and optionally a slice of
+    extra caveats that will be added to the discharge macaroon.
+    The caveatId parameter holds the still-encoded id of the caveat.
+    If the caveat kind was not recognised, the checker should return an
+    error with a ErrCaveatNotRecognized cause.
+    '''
+    __metaclass__ = abc.ABCMeta
+
+    @abc.abstractmethod
+    def check_third_party_caveat(self, ctx, info):
```

Comment:
This should have a doc comment that documents its argument and return types.

Reply:
done.